Mathlib/RingTheory/MvPowerSeries/PiTopology.lean

```diff
@@ -0,0 +1,243 @@
+
+theorem variables_tendsto_zero [Semiring R] :
+    Filter.Tendsto (X · : σ → MvPowerSeries σ R) Filter.cofinite (nhds 0) := by
+  classical
+  rw [tendsto_pi_nhds]
+  simp_rw [apply_eq_coeff]
+  intro d s hs
+  replace hs : 0 ∈ s := by simpa only [map_zero] using mem_of_mem_nhds hs
+  simp_rw [Filter.mem_map, Filter.mem_cofinite, ← Set.preimage_compl, coeff_X]
+  by_cases h : ∃ i, d = Finsupp.single i 1
+  · obtain ⟨i, rfl⟩ := h
+    apply (Set.finite_singleton i).subset
+    intro x
+    by_cases hx : x = i <;>
+      simp [Finsupp.single_eq_single_iff, Ne.symm, hx, hs]
+  · convert Set.finite_empty
+    rw [Set.eq_empty_iff_forall_not_mem]
+    intro x
+    rwa [Set.mem_preimage, Set.not_mem_compl_iff, if_neg]
+    tauto
+
+theorem tendsto_pow_zero_of_constantCoeff_nilpotent [CommSemiring R]
+    {f} (hf : IsNilpotent (constantCoeff σ R f)) :
+    Filter.Tendsto (fun n : ℕ => f ^ n) Filter.atTop (nhds 0) := by
+  classical
+  obtain ⟨m, hm⟩ := hf
+  simp_rw [tendsto_iff_coeff_tendsto, coeff_zero]
+  exact fun d ↦ tendsto_atTop_of_eventually_const fun n hn ↦
+    coeff_eq_zero_of_constantCoeff_nilpotent hm hn
+
+theorem tendsto_pow_zero_of_constantCoeff_zero [CommSemiring R]
+    {f} (hf : constantCoeff σ R f = 0) :
+    Filter.Tendsto (fun n : ℕ => f ^ n) Filter.atTop (nhds 0) := by
+  apply tendsto_pow_zero_of_constantCoeff_nilpotent
+  rw [hf]
+  exact IsNilpotent.zero
+
+/-- The powers of a `MvPowerSeries` converge to 0 iff its constant coefficient is nilpotent.
+N. Bourbaki, *Algebra II*, [bourbaki1981] (chap. 4, §4, n°2, corollaire de la prop. 3) -/
+theorem tendsto_pow_of_constantCoeff_nilpotent_iff [CommRing R] [DiscreteTopology R] (f) :
+    Filter.Tendsto (fun n : ℕ => f ^ n) Filter.atTop (nhds 0) ↔
+      IsNilpotent (constantCoeff σ R f) := by
+  refine ⟨?_, tendsto_pow_zero_of_constantCoeff_nilpotent⟩
+  intro h
+  suffices Filter.Tendsto (fun n : ℕ => constantCoeff σ R (f ^ n)) Filter.atTop (nhds 0) by
+    simp only [Filter.tendsto_def] at this
+    specialize this {0} _
+    suffices ∀ x : R, {x} ∈ nhds x by exact this 0
+    rw [← discreteTopology_iff_singleton_mem_nhds]; infer_instance
+    simp only [map_pow, Filter.mem_atTop_sets, ge_iff_le, Set.mem_preimage,
+      Set.mem_singleton_iff] at this
+    obtain ⟨m, hm⟩ := this
+    use m
+    apply hm m (le_refl m)
+  simp only [← @comp_apply _ R ℕ, ← Filter.tendsto_map'_iff]
+  simp only [Filter.Tendsto, Filter.map_le_iff_le_comap] at h ⊢
+  refine le_trans h (Filter.comap_mono ?_)
+  rw [← Filter.map_le_iff_le_comap]
+  exact Continuous.continuousAt (continuous_constantCoeff R)
+
+variable [Semiring R]
+
+/-- A power series is the sum (in the sense of summable families) of its monomials -/
+theorem hasSum_of_monomials_self (f : MvPowerSeries σ R) :
+    HasSum (fun d : σ →₀ ℕ => monomial R d (coeff R d f)) f := by
+  rw [Pi.hasSum]
+  intro d
+  convert hasSum_single d ?_ using 1
+  exact (coeff_monomial_same d _).symm
+  exact fun d' h ↦ coeff_monomial_ne (Ne.symm h) _
```

Comment:
Please use focussing dots (or restructure the proof to avoid having multiple goals)